rust/worker/src/server.rs

```diff
@@ -71,28 +84,75 @@ impl Configurable<QueryServiceConfig> for WorkerServer {
         // TODO: inject blockfile provider somehow
         // TODO: real path
         let path = PathBuf::from("~/tmp");
-        Ok(WorkerServer {
+        Ok(QueryServer {
             dispatcher: None,
             system: None,
             sysdb,
             log,
             hnsw_index_provider: HnswIndexProvider::new(storage.clone(), path),
             blockfile_provider: BlockfileProvider::new_arrow(storage),
             port: config.my_port,
+            memberlist: None,
+            server_state: QueryServerState::Pending,
         })
     }
 }
 
-impl WorkerServer {
-    pub(crate) async fn run(worker: WorkerServer) -> Result<(), Box<dyn std::error::Error>> {
-        let addr = format!("[::]:{}", worker.port).parse().unwrap();
-        println!("Worker listening on {}", addr);
+// ============== Component Implementation ==============
+#[async_trait]
+impl Component for QueryServer {
+    fn get_name() -> &'static str {
+        "QueryServer"
+    }
+
+    fn queue_size(&self) -> usize {
+        1000
+    }
+
+    async fn on_start(&mut self, _ctx: &crate::system::ComponentContext<Self>) -> () {}
+}
+
+impl Debug for QueryServer {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "QueryServer")
+    }
+}
+
+#[async_trait]
+impl Handler<Memberlist> for QueryServer {
+    async fn handle(&mut self, message: Memberlist, _ctx: &ComponentContext<QueryServer>) {
+        // TODO: for now we just overwrite the memberlist, we will need to check
+        // the version and only update if the version is newer
+        self.memberlist = Some(message);
+
+        let self_clone = self.clone();
```

Comment:
Why do we have to clone self?

Reply:
Rust will not be happy otherwise: `self` is a reference that is only valid in the method body.

Reply:
Yeah I see, async_move below, this makes the code structure a bit funky since we have a dangling server that will theoretically never get dropped